### PR TITLE
Add Qwen3-Omni moe thinker

### PR DESCRIPTION
**Pull request by wangxiongts**
This PR from the Qwen team for:  qwen3-omni-moe thinker part.

Testing has been conducted internally across four configurations (v0/v1, eager/CUDA) on several representative benchmarks, with results meeting expectations.

Known issues (we hope to resolve them together with the vLLM team):

- In v1 mode, use_audio_in_video will raise errors because the video mm_data and placeholders is not updated.

We sincerely appreciate the great work and support from the vLLM team, and look forward to your feedback.

CLOSE https://github.com/vllm-project/vllm/issues/25472

### CODE REVIEW

**Review by DarkLight1337 — COMMENTED** [review]
Thanks, can you update `tests/models/registry.py` to be able to pass the CI?

Also please update the Supported Models page

**Comment by wangxiongts** [comment]
Alright, I'll handle these parts. Currently, I'm still working on adding audio-in-video support in v1, In the meantime, One known issue is that I may not be able to straightforwardly reuse relevant modules from Qwen3-VL, because our model has already been made public, and some checkpoint keys and configurations are incompatible with Qwen3-VL. This stems from the fact that our internal iterations were not synchronized. This issue may require further careful discussion.

I might go on vacation starting tomorrow and probably won't resume modifications until after October 4th :)   You can proceed with the review based on the current version.

File: vllm/model_executor/layers/rotary_embedding/mrope.py [hunk 1]
```diff
@@ -1060,7 +1268,27 @@ def _omni_get_input_positions_tensor(
...
+        
+        if "qwen3_omni" in model_type:
+            llm_positions, mrope_position_delta = cls._omni3_get_input_positions_tensor(thinker_config, 
+                                                                                        torch.tensor([input_tokens]), 
+                                                                                        image_grid_thw, 
+                                                                                        video_grid_thw, 
+                                                                                        None, 
+                                                                                        use_audio_in_video, 
+                                                                                        audio_feature_lengths, 
+                                                                                        torch.tensor([1] * torch.tensor(video_grid_thw).shape[0]))
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/rotary_embedding/mrope.py]:
```suggestion
                                                                                        torch.ones(len(video_grid_thw))
```

Simplify this

File: vllm/model_executor/models/qwen3_omni_moe_thinker.py [hunk 2]
```diff
@@ -0,0 +1,1218 @@
...
+    def fast_pos_embed_interpolate(self, grid_thw):
+        num_grid_per_side = self.image_size // self.patch_size
+
+        idx_list = [[] for _ in range(4)]
+        weight_list = [[] for _ in range(4)]
+
+        # TODO: use torch instand of np
+        for t, h, w in grid_thw:
+            h_idxs = np.linspace(0, num_grid_per_side-1, h)
+            w_idxs = np.linspace(0, num_grid_per_side-1, w)
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/qwen3_omni_moe_thinker.py]: Are you able to finish this TODO before you have to go OOO?

**Comment by CHNtentes** [comment]
Thanks for your work. May I ask, will talker model get supported in future? It seems Qwen2.5-Omni still only support thinker model now.

**Comment by Wesley-Jzy** [comment]
LGTM! May I know whether Talker model will be supported by vLLM?

**Comment by ywang96** [comment]
Supporting Qwen3-Omni end-to-end will not be within the scope of `vllm-project/vllm`, but we already have some plan on supporting this model under a different project but leveraging the thinker models support from vLLM. Stay tuned!

**Comment by Wesley-Jzy** [comment]
> Supporting Qwen3-Omni end-to-end will not be within the scope of `vllm-project/vllm`, but we already have some plan on supporting this model under a different project but leveraging the thinker models support from vLLM. Stay tuned!

So that means vLLM project will support thinking model which is just like normal LLM model. And a new multimodal inference project will support end2end Qwen3-Omni model? Do I know more about this new project?

**Comment by ywang96** [comment]
> > Supporting Qwen3-Omni end-to-end will not be within the scope of `vllm-project/vllm`, but we already have some plan on supporting this model under a different project but leveraging the thinker models support from vLLM. Stay tuned!
> 
> So that means vLLM project will support thinking model which is just like normal LLM model. And a new multimodal inference project will support end2end Qwen3-Omni model? Do I know more about this new project?

Yea that's the right understanding! We're still planning for the new project so stay tuned!

**Comment by Wesley-Jzy** [comment]
> > > Supporting Qwen3-Omni end-to-end will not be within the scope of `vllm-project/vllm`, but we already have some plan on supporting this model under a different project but leveraging the thinker models support from vLLM. Stay tuned!
> > 
> > 
> > So that means vLLM project will support thinking model which is just like normal LLM model. And a new multimodal inference project will support end2end Qwen3-Omni model? Do I know more about this new project?
> 
> Yea that's the right understanding! We're still planning for the new project so stay tuned!

Great! And may I know will the new project also handle the single-model multimodal models such as Kimi-Audio? Or they will be supported by vLLM?

**Comment by CHNtentes** [comment]
> > > Supporting Qwen3-Omni end-to-end will not be within the scope of `vllm-project/vllm`, but we already have some plan on supporting this model under a different project but leveraging the thinker models support from vLLM. Stay tuned!
> > 
> > 
> > So that means vLLM project will support thinking model which is just like normal LLM model. And a new multimodal inference project will support end2end Qwen3-Omni model? Do I know more about this new project?
> 
> Yea that's the right understanding! We're still planning for the new project so stay tuned!

Really wish the new project is fast and efficient. Tried transformers and audio output was SLOW...

**Comment by eschmidbauer** [comment]
> Really wish the new project is fast and efficient. Tried transformers and audio output was SLOW...

Same, even with flash-attn2 it is very slow

**Comment by CHNtentes** [comment]
> > Really wish the new project is fast and efficient. Tried transformers and audio output was SLOW...
> 
> Same, even with flash-attn2 it is very slow

I tried this PR and it's like >20x faster than transformers :)

File: vllm/model_executor/layers/rotary_embedding/mrope.py [hunk 3]
```diff
@@ -1028,6 +1028,214 @@ def _vl_get_input_positions_tensor(
...
+    def _omni3_get_input_positions_tensor(
+        cls,
+        config,
+        input_ids: Optional[torch.LongTensor] = None,
+        image_grid_thw: Optional[torch.LongTensor] = None,
+        video_grid_thw: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        use_audio_in_video: bool = False,
+        audio_seqlens: Optional[torch.LongTensor] = None,
+        second_per_grids: Optional[torch.Tensor] = None,
```
Comment (ywang96) [inline vllm/model_executor/layers/rotary_embedding/mrope.py:1131]: We should simplify this implementation at least - it looks like currently we have an additional unused batch dimension of `input_ids` as well as an unnecessary attention mask. 

I assumed this was adopted from the training code, so I will clean it up.

**Comment by UmutAlihan** [comment]
Looking very forward to be able to inference using vLLM for this omni model. Compiling vllm from specific branch source is really hustling on some limited access corporate standalone servers 😿.

**Comment by tensorboy** [comment]
> FYI I'm back to working on this PR

you are the best!

**Comment by ywang96** [comment]
This PR should be again functional - I probably won't fix the `audio_in_video` issue and will have it fixed together in #26156 so should be able to merge it pretty soon

**Comment by ywang96** [comment]
> Thank you for all the hard work. We really appreciate it.
> 
> From what I understand this PR is only for the thinker model and it only supports text output.
> 
> Is there any rough timeline when audio output will be added? Thank you so much!

We don't have a rough timeline yet but hopefully by the end of this year/early next year

**Review by ywang96 — APPROVED** [review]

**Review by DarkLight1337 — APPROVED** [review]
Thanks for your hard work!

**Review by Isotr0py — APPROVED** [review]
The processor tests should pass now:
```
tests/models/multimodal/processing/test_common.py::test_processing_correctness[1.0-32-0.3-Qwen/Qwen3-Omni-30B-A3B-Instruct] Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'mrope_interleaved', 'interleaved', 'mrope_section'}
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'interleaved', 'mrope_section'}
INFO 10-10 23:00:59 [model.py:653] Resolved architecture: Qwen3OmniMoeForConditionalGeneration
`torch_dtype` is deprecated! Use `dtype` instead!
INFO 10-10 23:00:59 [model.py:1714] Using max model len 65536
PASSED
tests/models/multimodal/processing/test_common.py::test_processing_correctness[1.0-32-0.5-Qwen/Qwen3-Omni-30B-A3B-Instruct] Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'mrope_interleaved', 'interleaved', 'mrope_section'}
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'interleaved', 'mrope_section'}
INFO 10-10 23:01:17 [model.py:653] Resolved architecture: Qwen3OmniMoeForConditionalGeneration
INFO 10-10 23:01:17 [model.py:1714] Using max model len 65536
PASSED
tests/models/multimodal/processing/test_common.py::test_processing_correctness[1.0-32-1.0-Qwen/Qwen3-Omni-30B-A3B-Instruct] Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'mrope_interleaved', 'interleaved', 'mrope_section'}
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'interleaved', 'mrope_section'}
INFO 10-10 23:01:26 [model.py:653] Resolved architecture: Qwen3OmniMoeForConditionalGeneration
INFO 10-10 23:01:26 [model.py:1714] Using max model len 65536
PASSED
```
And outputs look reasonable on my side too:
```
Adding requests: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.86s/it]
Processed prompts: 100%|██████████████████████████████████████████████████████████████████████| 1/1 [00:13<00:00, 13.77s/it, est. speed input: 298.63 toks/s, output: 22.00 toks/s]
Based on the audio, video, and image content, here is a breakdown of what is happening:

### Audio Content

The audio features a man reciting a well-known nursery rhyme, "Mary Had a Little Lamb." The rhyme is:

> "Mary had a little lamb, its fleece was white as snow. And everywhere that Mary went, the lamb was sure to go."

The speaker mentions that these were the "first words" spoken into a phonograph, which is a historical reference to the invention of the phonograph by Thomas Edison.

### Image Content

The image shows a young child, likely a toddler, sitting on a bed. The child is wearing glasses and is looking at a book. The child's feet are propped up on the book, and they are turning the pages. The child appears to be enjoying the book and is smiling.

### Why the Video is Funny

The humor in the video comes from the contrast between the child's serious and focused demeanor and the absurdity of the situation. The child is wearing glasses and is sitting on the bed with their feet propped up on the book, which is a common and relatable scene for many people. However, the child's intense focus on the book, combined with the glasses, creates a humorous and endearing image. The child seems to be taking the reading experience very seriously, which is funny because they are just a toddler. The overall effect is a charming and amusing scene that many viewers can relate to and find funny.
```
